src/python_tools/plotting/plot_pass_generator.py

```diff
@@ -0,0 +1,116 @@
+from bokeh.plotting import Figure
+from bokeh.models import ColumnDataSource
+from bokeh.models.tools import HoverTool
+import numpy as np
+
+
+class PassGeneratorPlotter:
+    def plot_zones(self, zones: list):
+        """
+        Plots pass generator zones from the Python bindings as per
+        software/python_bindings/pass_generator.cpp.
+        :param zones: A list of dicts describing rectangular passing zones.
+        """
+        zones_dict = dict(center_xs=[], center_ys=[], widths=[], heights=[])
+
+        for zone in zones:
+            zones_dict["center_xs"].append(zone.centre().x())
+            zones_dict["center_ys"].append(zone.centre().y())
+            zones_dict["widths"].append(zone.xLength())
+            zones_dict["heights"].append(zone.yLength())
+
+        self.zones_source.data.update(zones_dict)
+
+    def plot_passes(self, passes: list):
+        """
+        Plots dicts describing passes from Python bindings as per
+        software/python_bindings/pass_utilities.cpp.
+        :param passes: A list of dicts describing passes and their associated ratings.
+        """
+        passes_dict = dict(
+            receiver_xs=[],
+            receiver_ys=[],
+            pass_line_xs=[],
+            pass_line_ys=[],
+            pass_rating=[],
+            adjusted_pass_rating=[],
+        )
+
+        for pass_and_rating_dict in passes:
+            the_pass = pass_and_rating_dict["pass"]
+            rating = pass_and_rating_dict["rating"]
+
+            passer_point = the_pass["passer_point"]
+            receiver_point = the_pass["receiver_point"]
+
+            passes_dict["receiver_xs"].append(receiver_point.x())
+            passes_dict["receiver_ys"].append(receiver_point.y())
+
+            passes_dict["pass_line_xs"].append([passer_point.x(), receiver_point.x()])
+            passes_dict["pass_line_ys"].append([passer_point.y(), receiver_point.y()])
+
+            passes_dict["pass_rating"].append(rating)
+            # adjusted pass rating is only used to determine the alpha of the
+            # line segment between passer and receiver points
+            passes_dict["adjusted_pass_rating"].append(np.clip(rating * 2, 0, 1))
```

Comment:
Changing the alpha makes bad passes hard to see, right? Could we change the brightness instead, i.e. dark orange to light orange.

Reply:
made some visualization changes as below